crates/proto/src/h3/h3_server.rs

```diff
@@ -0,0 +1,127 @@
+
+impl H3Server {
+    /// Construct the new Acceptor with the associated pkcs12 data
+    pub async fn new(
+        name_server: SocketAddr,
+        cert: Vec<Certificate>,
+        key: PrivateKey,
+    ) -> Result<Self, ProtoError> {
+        // setup a new socket for the server to use
+        let socket = <tokio::net::UdpSocket as UdpSocket>::bind(name_server).await?;
+        Self::with_socket(socket, cert, key)
+    }
+
+    /// Construct the new server with an existing socket
+    pub fn with_socket(
+        socket: tokio::net::UdpSocket,
+        cert: Vec<Certificate>,
+        key: PrivateKey,
+    ) -> Result<Self, ProtoError> {
+        let mut config = TlsServerConfig::builder()
+            .with_safe_default_cipher_suites()
+            .with_safe_default_kx_groups()
+            .with_protocol_versions(&[&TLS13])
+            .expect("TLS1.3 not supported")
+            .with_no_client_auth()
+            .with_single_cert(cert, key)?;
+
+        config.alpn_protocols = vec![ALPN_H3.to_vec()];
+
+        let mut server_config = ServerConfig::with_crypto(Arc::new(config));
+        server_config.transport = Arc::new(super::transport());
+
+        let socket = socket.into_std()?;
+
+        let endpoint = Endpoint::new(
+            EndpointConfig::default(),
+            Some(server_config),
+            socket,
+            Arc::new(quinn::TokioRuntime),
+        )?;
+
+        Ok(Self { endpoint })
+    }
+
+    /// Accept the next incoming connection.
+    ///
+    /// # Returns
+    ///
+    /// A remote connection that could accept many potential requests and the remote socket address
+    pub async fn accept(&mut self) -> Result<Option<(H3Connection, SocketAddr)>, ProtoError> {
+        let connecting = match self.endpoint.accept().await {
+            Some(conn) => conn,
+            None => return Ok(None),
+        };
+
+        let remote_addr = connecting.remote_address();
+        let connection = connecting.await?;
+        Ok(Some((
+            H3Connection {
+                connection: Connection::new(h3_quinn::Connection::new(connection))
+                    .await
+                    .map_err(|e| ProtoError::from(format!("h3 connection failed: {e}")))?,
+            },
+            remote_addr,
+        )))
+    }
+
+    /// Returns the address this server is listening on
+    ///
+    /// This can be useful in tests, where a random port can be associated with the server by binding on `127.0.0.1:0` and then getting the
+    ///   associated port address with this function.
+    pub fn local_addr(&self) -> Result<SocketAddr, io::Error> {
+        self.endpoint.local_addr()
+    }
+}
+
+/// A HTTP/3 connection.
+pub struct H3Connection {
+    connection: Connection<h3_quinn::Connection, Bytes>,
+}
+
+impl H3Connection {
+    /// Accept the next request from the client
+    pub async fn accept(
+        &mut self,
+    ) -> Option<Result<(Request<()>, RequestStream<BidiStream<Bytes>, Bytes>), ProtoError>> {
```

Comment:
Should I also wrap the returned types here? Otherwise they require `h3` and `h3-quinn` to be used.

Reply:
The main benefit I see is if it simplifies working with those streams to wrap them in a type more specific to DNS operations...

Reply:
It wouldn't, I was only trying avoiding exposing public types. Our wrapper would have basically the same API we are using now already.